components/raftstore-v2/src/operation/command/admin/split.rs

```diff
@@ -0,0 +1,884 @@
+        // following PR.
+        let region_id = derived.get_id();
+        let tablet = self.tablet().unwrap();
+
+        for new_region in &regions {
+            let new_region_id = new_region.id;
+            if new_region_id == region_id {
+                continue;
+            }
+
+            let new_tablet_path = self
+                .tablet_factory()
+                .split_tablet_path(new_region_id, RAFT_INIT_LOG_INDEX);
+            tablet
+                .create_checkpoint(&new_tablet_path)
+                .unwrap_or_else(|e| {
+                    panic!(
+                        "{:?} fails to create checkpoint with path {:?}: {:?}",
+                        self.logger.list(),
+                        new_tablet_path,
+                        e
+                    )
+                });
+        }
+
+        let new_tablet_path = self
+            .tablet_factory()
+            .split_tablet_path(region_id, log_index);
+        // Change the tablet path by using the new tablet suffix
+        tablet
+            .create_checkpoint(&new_tablet_path)
+            .unwrap_or_else(|e| {
+                panic!(
+                    "{:?} fails to create checkpoint with path {:?}: {:?}",
+                    self.logger.list(),
+                    new_tablet_path,
+                    e
+                )
+            });
+        // Here, we open the tablet without registering as it is the split version. We
+        // will register it when switching it to the normal version.
+        let tablet = self
+            .tablet_factory()
+            .open_tablet(
+                region_id,
+                Some(log_index),
+                OpenOptions::default()
+                    .set_create(true)
+                    .set_split_use(true)
+                    .set_skip_cache(true),
+            )
+            .unwrap();
+        self.publish_tablet(tablet);
+
+        // Clear the write batch belonging to the old tablet
+        self.clear_write_batch();
+
+        self.region_state_mut().set_region(derived.clone());
+        self.region_state_mut().set_tablet_index(log_index);
+        // self.metrics.size_diff_hint = 0;
+        // self.metrics.delete_keys_hint = 0;
+
+        let mut resp = AdminResponse::default();
+        resp.mut_splits().set_regions(regions.clone().into());
+        PEER_ADMIN_CMD_COUNTER.batch_split.success.inc();
+
+        Ok((
+            resp,
+            AdminCmdResult::SplitRegion(SplitResult {
+                regions,
+                derived,
+                tablet_index: log_index,
+            }),
+        ))
+    }
+}
+
+fn get_range_not_in_region(region: &metapb::Region) -> Vec<Range<'_>> {
+    let mut ranges = Vec::new();
+    if !region.get_start_key().is_empty() {
+        ranges.push(Range::new(b"", region.get_start_key()));
+    }
+    if !region.get_end_key().is_empty() {
+        ranges.push(Range::new(region.get_end_key(), b""));
```

Comment:
are you sure here the end key can be "" and the caller will handle it properly?
In v1 it's converted to  keys::DATA_MAX_KEY